feat(formatter): implement todos in JsAnyExpression

Boshen opened the pull request:
## Summary
This PR implements most of the todos in `rome_formatter/src/ts/expressions/expression.rs`
And fixes:
* incorrect `import.meta` syntax
* incorrect `/=` `**=` for JsAssignmentExpression

This PR is rebased onto https://github.com/rome/tools/pull/1843 for passing tests for `a >> b`
## Test Plan
See updated formatter tests

Boshen commented:
btw I can't figure out what syntax `TsAssertion` and `TsConstAssertion` is 😿 , rslint gives no hint either
```
TsAssertion = expr:JsAnyExpression ident: Ident '<' ty:TsType '>'
TsConstAssertion = expr:JsAnyExpression ident: Ident '<' 'const' '>'
```

MichaReiser commented on crates/rome_formatter/src/ts/expressions/any_reference_member.rs:
```diff
@@ -0,0 +1,29 @@
+
+impl ToFormatElement for JsReferencePrivateMember {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		formatter.format_token(&self.name_token()?)
```
Private names have the form `#name` where the `#` is its own token inside private member. So you need to visit (format) the `#` token too to make sure it doesn't get lost. I'm a bit surprised that the test is passing

Boshen replied:
fixed in e311e73, although the parser hasn't implemented this yet :-( (which is why the test looked strange ... removed the test for now)

MichaReiser commented on crates/rslint_parser/test_data/inline/ok/import_meta.rast:
```diff
@@ -6,6 +6,7 @@ JsRoot {
             expression: ImportMeta {
                 import_token: IMPORT_KW@0..6 "import" [] [],
                 dot_token: DOT@6..7 "." [] [],
+                meta_token: missing (required),
```
Seems like the parser isn't parsing the meta token. I'll be working on `import` soon. Should we defer adding the meta token until then?

Boshen replied:
We can leave it here and fix later I think.

MichaReiser replied:
We shouldn't have any parsing errors in any file inside of the `ok` folder and this is a parsing error. The easiest will be to just remove the whole grammar for now since it isn't supported anyway.

Boshen replied:
> We shouldn't have any parsing errors in any file inside of the ok folder and this is a parsing error. The easiest will be to just remove the whole grammar for now since it isn't supported anyway.

I have reverted the change.

MichaReiser commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +17,362 @@ impl ToFormatElement for JsAnyExpression {
+
+impl ToFormatElement for JsUnaryExpression {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		// TODO handle `+` `-` `~` `!`
```
Why do they need special handling?

Boshen replied:
We need to handle spacing for `+` `-` `~` `!` vs `void` `delete` `typeof`, fixed in 1936dd0

MichaReiser commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +17,362 @@ impl ToFormatElement for JsAnyExpression {
+
+impl ToFormatElement for JsBinaryExpression {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		let right = if let Some(token) = self.right() {
+			formatter.format_node(token)?
+		} else {
+			empty_element()
+		};
```
Oh, the return-type of `right` on `BinaryExpression` is incorrect. `right` should return a `SyntaxResult<JsAnyExpression>` Would you mind changing the implementation?

Boshen replied:
fixed in 0b01cd5

MichaReiser commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +17,362 @@ impl ToFormatElement for JsAnyExpression {
+
+impl ToFormatElement for JsConditionalExpression {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		Ok(format_elements![
+			formatter.format_node(self.test()?)?,
+			space_token(),
+			formatter.format_token(&self.question_mark_token()?)?,
+			space_token(),
+			formatter.format_node(self.consequent()?)?,
+			space_token(),
+			formatter.format_token(&self.colon_token()?)?,
+			space_token(),
+			formatter.format_node(self.alternate()?)?,
+		])
```
Can you add a test case that exceeds the 80 character line width?  We then want to wrap the condition over multiple lines. Similar to rome classic

https://github.com/rome/tools/blob/c76ccc0345fe00ea4b3f684424065d05c3ab3abe/internal/formatter/builders/js/expressions/JSConditionalExpression.ts#L48-L68

MichaReiser commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +17,362 @@ impl ToFormatElement for JsAnyExpression {
+		} else {
+			empty_element()
+		};
+		Ok(format_elements![
+			formatter.format_node(self.left()?)?,
+			space_token(),
+			formatter.format_token(&self.operator()?)?,
+			space_token(),
+			right
+		])
+	}
```
This is fine for now but we probably want to print binary expressions across multiple lines if they get very long, similar to what rome classic

https://github.com/rome/tools/blob/8edbed0c3505c8ed3724015df83817c2a92acce8/internal/formatter/builders/js/expressions/JSBinaryExpression.ts#L19-L81

MichaReiser commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +17,362 @@ impl ToFormatElement for JsAnyExpression {
+}
+
+impl ToFormatElement for JsAssignmentExpression {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		Ok(format_elements![
+			formatter.format_node(self.left()?)?,
+			space_token(),
+			formatter.format_token(&self.operator_token()?)?,
+			space_token(),
+			formatter.format_node(self.right()?)?,
+		])
+	}
```
We may want to make sure that assignments can break if they're very long (or concatenated)

https://github.com/rome/tools/blob/8dcd9dde2b0c3d850f24279c472e0e2fd90bb0af/internal/formatter/builders/js/utils.ts#L271-L285

MichaReiser commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +17,362 @@ impl ToFormatElement for JsAnyExpression {
+	}
+}
+
+impl ToFormatElement for ImportMeta {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		Ok(format_elements![
+			formatter.format_token(&self.import_token()?)?,
+			formatter.format_token(&self.dot_token()?)?,
+			formatter.format_token(&self.meta_token()?)?,
+		])
+	}
+}
```
I just checked the grammar and this seems to be a dead node. The `ImportMeta` node is never used on the right-hand side of any other rule. I think we can delete this.

ematipico replied:
It's worth moving this inside a `import/meta.rs` file

MichaReiser commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +17,362 @@ impl ToFormatElement for JsAnyExpression {
+		} else {
+			empty_element()
+		};
+		Ok(format_elements![
+			formatter.format_node(self.left()?)?,
+			space_token(),
+			formatter.format_token(&self.operator()?)?,
+			space_token(),
+			right,
+		])
+	}
```
Same as for binary expression, we probably want to indent the right hand side if the expressions are very long

MichaReiser commented on crates/rome_formatter/tests/specs/js/conditional.js:
```diff
@@ -0,0 +1,4 @@
+a  ?  b  :  c
+d
+	? (e + f)
+	: (g + h)
```
I think it should be fine to print these on a single line because the whole expression isn't exceeding the line width. 

Can we add another example for a conditional that exceeds 80 character line width? Maybe even with nesting?

ematipico commented:
> btw I can't figure out what syntax `TsAssertion` and `TsConstAssertion` is 😿 , rslint gives no hint either
> 
> ```
> TsAssertion = expr:JsAnyExpression ident: Ident '<' ty:TsType '>'
> TsConstAssertion = expr:JsAnyExpression ident: Ident '<' 'const' '>'
> ```

You ignore anything related to TypeScript at the moment. We need to review it and probably it will change drastically.

MichaReiser commented on crates/rome_formatter/tests/specs/js/expression.js:
```diff
@@ -0,0 +1,8 @@
+x = y!
+x = ++y
+x = --y
+x = y++
+x = y--
+x ?? y
+x || y
+x && y
```
Can we add some examples that exceed the line width? Maybe even with nesting?

MichaReiser commented on crates/rome_formatter/tests/specs/js/static_member.expected.js:
```diff
@@ -0,0 +1,3 @@
+a.b;
+c?.d;
+e.#f
```
Shouldn't there be a `;` too?
```suggestion
e.#f;
```

Boshen replied:
Hah! I thought this was a bug somewhere, and then I printed the diagnostics and saw ”private identifiers are unsupported“. I'll remove this test for now.

MichaReiser replied:
Should work fine after  #1851 landed

MichaReiser reviewed (commented):
This is amazing, thank you!

I believe private names aren't printed correctly yet. 

There's also the question how far you wanna go. There are some expressions that should break if exceeding the line width but I'm fine if we leave this as is for now and tackle it in later PRs.

Boshen commented:
@MichaReiser All actionable items are done, we can work on the finer details later, let's put more focus on the parser for now 😀

Boshen commented on crates/rome_formatter/tests/specs/js/static_member.expected.js:
```diff
@@ -0,0 +1,2 @@
+a.b;
+c?.d;
```
"private identifiers are unsupported" in the parser 😢 
https://github.com/rome/tools/blob/main/crates/rslint_parser/src/syntax/expr.rs#L531

I should probably look into outputting all the error diagnostics from the parser when running formatter tests, the formatter currently can give the wrong result which looks correct...

ematipico replied:
Makes sense, you have to create a snippet with a class:

```js
class B {
	foo() { #a.b }
}
```

etc.

Boshen replied:
Shall we skip `#` for now? I played around a bit and the parser just breaks when it sees `#`

MichaReiser replied:
#1851  enables class field support

ematipico commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +18,363 @@ impl ToFormatElement for JsAnyExpression {
+	}
+}
+
+impl ToFormatElement for JsClassExpression {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		let extends = if let Some(extends_clause) = self.extends_clause() {
+			format_elements![space_token(), formatter.format_node(extends_clause)?]
+		} else {
+			empty_element()
+		};
+
+		let id = if let Some(token) = self.id() {
+			format_elements![space_token(), formatter.format_node(token)?]
+		} else {
+			empty_element()
+		};
+
+		Ok(format_elements![
+			formatter.format_token(&self.class_token()?)?,
+			id,
+			extends,
+			space_token(),
+			group_elements(format_elements![
+				formatter.format_token(&self.l_curly_token()?)?,
+				block_indent(join_elements(
+					hard_line_break(),
+					formatter.format_nodes(self.members())?
+				)),
+				formatter.format_token(&self.r_curly_token()?)?
+			])
+		])
+	}
+}
```
Would you mind moving this inside a `class.rs` file? Ideally we should move there all formatters related to `class`

ematipico commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +18,363 @@ impl ToFormatElement for JsAnyExpression {
+	}
+}
+
+impl ToFormatElement for JsStaticMemberExpression {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		Ok(format_elements![
+			formatter.format_node(self.object()?)?,
+			formatter.format_token(&self.operator()?)?,
+			formatter.format_node(self.member()?)?,
+		])
+	}
+}
```
To move to a `class.rs` file

ematipico commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +18,363 @@ impl ToFormatElement for JsAnyExpression {
 	}
 }
+
+impl ToFormatElement for Template {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		// TODO implement template string, see parser in rslint_parser/src/ast/expr_ext.rs
+		Ok(format_elements![
+			formatter.format_token(&self.backtick_token()?)?,
+			token("TODO"),
+			formatter.format_token(&self.backtick_token()?)?
+		])
+	}
+}
```
Let's move it into a `template.rs` file. This will change and probably might have some custom logic

ematipico commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +18,363 @@ impl ToFormatElement for JsAnyExpression {
+	}
+}
+
+impl ToFormatElement for JsFunctionExpression {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		let mut tokens = vec![];
+
+		if let Some(token) = self.async_token() {
+			tokens.push(formatter.format_token(&token)?);
+			tokens.push(space_token());
+		}
+
+		tokens.push(formatter.format_token(&self.function_token()?)?);
+
+		if let Some(token) = self.star_token() {
+			tokens.push(formatter.format_token(&token)?);
+		}
+
+		tokens.push(space_token());
+		if let Some(token) = self.id() {
+			tokens.push(formatter.format_node(token)?);
+		}
+		tokens.push(formatter.format_node(self.parameters()?)?);
+		tokens.push(space_token());
+		tokens.push(formatter.format_node(self.body()?)?);
+
+		Ok(concat_elements(tokens))
+	}
+}
```
Let's move it inside a `function.rs` file. There we will have also `JsFunctionDeclaration`

ematipico commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +18,363 @@ impl ToFormatElement for JsAnyExpression {
+	}
+}
+
+impl ToFormatElement for JsImportCallExpression {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		Ok(format_elements![
+			formatter.format_token(&self.import_token()?)?,
+			formatter.format_token(&self.l_paren_token()?)?,
+			formatter.format_node(self.argument()?)?,
+			formatter.format_token(&self.r_paren_token()?)?,
+		])
+	}
+}
```
Inside a `import/mod.rs` file

ematipico commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +18,363 @@ impl ToFormatElement for JsAnyExpression {
+	}
+}
+
+impl ToFormatElement for TsNonNull {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		Ok(format_elements![
+			formatter.format_node(self.expr()?)?,
+			formatter.format_token(&self.excl_token()?)?,
+		])
+	}
+}
```
Can we just have a `todo!()` here?

Boshen replied:
done

ematipico commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +18,363 @@ impl ToFormatElement for JsAnyExpression {
+	}
+}
+
+impl ToFormatElement for JsPreUpdateExpression {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		Ok(format_elements![
+			formatter.format_token(&self.operator()?)?,
+			formatter.format_node(self.operand()?)?,
+		])
+	}
+}
+
+impl ToFormatElement for JsPostUpdateExpression {
+	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
+		Ok(format_elements![
+			formatter.format_node(self.operand()?)?,
+			formatter.format_token(&self.operator()?)?,
+		])
+	}
+}
```
We could move these inside a `update_expressions.rs` file. What do you think?

ematipico commented on crates/rome_formatter/tests/specs/js/class.expected.js:
```diff
@@ -14,3 +14,5 @@ class Foo extends Boar {
 	double_semicolon = [5, 3, 4];
 }
+x = class {};
+x = class foo extends Boar {};
```
Let's add a test with a veeeery long names, both the class name and the super class.

If you're not sure what to do, we use Prettier as a reference

ematipico commented on crates/rome_formatter/tests/specs/js/function.js:
```diff
@@ -0,0 +1,17 @@
+}
+function* Foo() {
+  yield
+  yield x
+  yield* x
```
Let's add a couple of tests where `argument` (`x`) is really long

ematipico commented on crates/rome_formatter/tests/specs/js/import.js:
```diff
@@ -0,0 +1,2 @@
+import(x)
```
Let's add a test with an argument that has a long name

ematipico commented on crates/rome_formatter/tests/specs/js/new.expected.js:
```diff
@@ -0,0 +1,2 @@
+new a();
+new b(x);
```
Let's add a test with an argument that has a long name. Same for function name

ematipico commented on crates/rome_formatter/tests/specs/js/parenthesized.js:
```diff
@@ -0,0 +1 @@
+(1)
```
Let's add a test with an argument that has a long name

ematipico commented on crates/rome_formatter/tests/specs/js/unary.js:
```diff
@@ -0,0 +1,7 @@
+delete  a
+void    b
+typeof  c
+x = +1
+x = -1
+x = ~1
+x = !1
```
Let's test with long names too

ematipico reviewed (changes requested):
Thank you for tackling this! :)

I would expand the test base with long names and see how the formatter behaves.

We use Prettier as reference, so use it in case you don't know how to format something!

Boshen commented:
@MichaReiser @ematipico 
Rebased and updated with e534c7a and f4d5a2e

The formatter tests need a lot more work after looking at the prettier tests,
I'm currently just using the tests for TDD and guiding me through the parser.

MichaReiser commented on crates/rome_formatter/src/ts/expressions/expression.rs:
```diff
@@ -8,44 +17,242 @@ impl ToFormatElement for JsAnyExpression {
+	}
+}
+
+impl ToFormatElement for TsNonNull {
+	fn to_format_element(&self, _formatter: &Formatter) -> FormatResult<FormatElement> {
+		todo!()
+	}
+}
```
Delete, empty implementations are misleading. It's better to get a compile error in that case

MichaReiser commented on crates/rome_formatter/tests/specs/js/module.expected.js:
```diff
@@ -0,0 +1 @@
+import.
```
This doesn't seem to do any formatting. Can we delete this test (and the format implementation?)

MichaReiser reviewed (commented):
Looks good. Would you mind rebasing your changes and address the last two comments?

Boshen commented:
@MichaReiser rebased and squashed, removed:
* Template
* TsNonNull
* JsStaticMemberExpression (this introduced JsAnyName which is also a todo)
